sarjinius iP

sarjinius opened the pull request:
# Virtue

> "You have power over your mind - not outside events. Realize this, and you will find strength." ― Marcus Aurelius [(source)](https://www.goodreads.com/quotes/190580-you-have-power-over-your-mind---not-outside-events)

Virtue frees your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to use is,
1. download it from [here](https://github.com/sarjinius/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
#
If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method: 
```java
public class Virtue {
    public static void main(String[] args) {
        Virtue virtue = new Virtue();
        virtue.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

sarjinius committed:
Greets the user

sarjinius committed:
Exits with goodbye message

sarjinius committed:
Adds horizontal lines

sarjinius committed:
Adds echo method

sarjinius committed:
Echoes commands until bye is input.

sarjinius committed:
Adds indentations

sarjinius committed:
Renames command to task, creates a task list, modifies echo method to add task to list

sarjinius committed:
Adds method that prints task list

sarjinius committed:
Creates a new Task class

sarjinius committed:
Adds mark, unmark, and toString methods to Task class

sarjinius committed:
Adds isCommandBye, isCommandList, and getTask methods.

sarjinius committed:
Adds isCommandMark, isCommandUnmark, and getIndex methods.

sarjinius committed:
Adds markTaskAsDone and markTaskAsNotDone methods

sarjinius committed:
Adds classes Todo, Deadline and Event that extend the Task class.

sarjinius committed:
Adds isCommandTodo, isCommandDeadline, isCommandEvent, and numTasks methods.

sarjinius committed:
Adds methods that gets the fields to be used in the Todo, Deadline, and Event constructors.

sarjinius committed:
Replaces addTask method with addTodo, addDeadline, addEvent methods, modifies control flow

sarjinius committed:
Updates input and expected scripts for automated testing

sarjinius committed:
Handles empty description errors and unknown command errors

sarjinius committed:
Adds support for deleting tasks from the list.

dedsecrattle commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,389 @@
+import java.util.Objects;
+import java.util.Scanner;
+
 public class Duke {
```
Different Classes can be placed in a separate file for better readability?
e.g Creating Task.java to store the Class `Task`

dedsecrattle commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,389 @@
+    // Checks if the last input is todo.
+    private boolean isCommandTodo(String command) {
+        String firstWord = getFirstWord(command);
+        return Objects.equals(firstWord, "todo");
```
reductant use of `Object.equals` method can be checked using `String` equality

dedsecrattle commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,389 @@
+    }
+
+    // Takes inputs from user until bye has been input.
+    private void takeInputsUntilBye() {
```
method name although is intuitive , but it can be made a bit more general e.g `userInput`

dedsecrattle reviewed (commented):
Overall , I Like the code logic , needs improvement in Code structuring

sarjinius committed:
Uses Enums for Command Types and moves different classes to new files

sarjinius committed:
Moves methods from Virtue class to other classes

sarjinius committed:
Initializes task list from hard drive when chatbot starts up.

sarjinius committed:
Writes task list into file after list is modified.

sarjinius committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Writes task list into file after list is modified.
  Initializes task list from hard drive when chatbot starts up.

RingoftheKing commented on src/main/Duke/virtue/VirtueFileReader.java:
```diff
@@ -0,0 +1,38 @@
+import java.io.IOException;
+import java.util.Scanner;
+
+public class VirtueFileReader {
```
It would be nice to explain what the file reader does for the application in a comment. Might be more important when in the future we deal with different files with more than one purpose

RingoftheKing commented on src/main/Duke/virtue/VirtueTaskList.java:
```diff
@@ -0,0 +1,168 @@
+import java.util.List;
+
+public class VirtueTaskList {
+    // The task list.
```
nit: this comment may either be removed for now, or you can make it more descriptive as needed in the future.

RingoftheKing reviewed (commented):
Your code is really neat and follows the Coding Standard well! Most Classes are easy to understand, GJ!

ananya21 commented on src/main/Duke/virtue/Command.java:
```diff
@@ -0,0 +1,135 @@
+
+public class Command {
+    // The available types of commands in Virtue.
+    public enum CommandType {
```
Using an enum to define CommandType was a good idea, since it groups together related constants, and makes the code more readable

ananya21 commented on src/main/Duke/virtue/Command.java:
```diff
@@ -0,0 +1,135 @@
+    protected String from;
+    protected String to;
+
+    public Command(String command) throws VirtueException {
```
having javadoc comments here will help with understanding of the code

ananya21 commented on src/main/Duke/virtue/Command.java:
```diff
@@ -0,0 +1,135 @@
+            }
+        }
+
+        throw new UnknownCommandTypeException();
```
using specific exceptions makes error handling more precise and helps with debugging

ananya21 commented on src/main/Duke/virtue/Virtue.java:
```diff
@@ -0,0 +1,81 @@
+    VirtueTaskList taskList = new VirtueTaskList();
+
+    // A horizontal line.
+    private static final String horizontalLine = "____________________________________________________________";
```
its very good that you made this final, and abstracted printing with indents and printing horizontal lines, it would have been very difficult to read otherwise

ananya21 commented on src/main/Duke/virtue/Virtue.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    // Runs the chatbot.
+    private void run() {
```
its really good you abstracted out the functions, and made the run() method concise

ananya21 commented on src/main/Duke/virtue/VirtueTaskList.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    // Adds a todo task to the task list.
+    private void addTodo(String description) {
```
Method names like addTodo, addDeadline, addEvent, markTaskAsDone, and printOut clearly convey their functionality, which is excellent for code readability.

ananya21 commented on src/main/Duke/virtue/VirtueTaskList.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    // Adds a deadline task to the task list.
+    private void addDeadline(String description, String by) {
```
the code for updating the user about task addition and completion is repeated across 'addTodo', 'addDeadline', 'addEvent'. It might be useful to abstract this code into a method that accepts parameters to customize the message

ananya21 reviewed (commented):
the use of specific exceptions, and using the enum data type and good use of OOP principles are strong aspects of your code, places where you could improve on is code duplication (simplifying your code) and placing Java documentation to make the code more readable and understandable

sarjinius committed:
Adds ability for chatbot to understand dates and times.

sarjinius committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Adds ability for chatbot to understand dates and times.

sarjinius committed:
Code refactored to add more OOP

sarjinius committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

sarjinius committed:
Can run the chatbot using Gradle

sarjinius committed:
TodoTest.java cannot import Todo class

sarjinius committed:
Added testing ability for Todo and VirtueTaskList classes.

sarjinius committed:
Changes .jar name created to virtue.jar

sarjinius committed:
Added JavaDoc to most classes and methods

sarjinius committed:
Code modified to follow the coding standard

sarjinius committed:
JavaDoc modified to follow coding standard

sarjinius committed:
Adds a find method that can filter the task list by a keyword.

sarjinius committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Adds a find method that can filter the task list by a keyword.

sarjinius committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  JavaDoc modified to follow coding standard
  Added JavaDoc to most classes and methods

# Conflicts:
#	src/main/java/virtue/VirtueTask.java
#	src/main/java/virtue/VirtueTaskList.java

sarjinius committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Code modified to follow the coding standard

# Conflicts:
#	src/main/java/virtue/Ui.java
#	src/main/java/virtue/Virtue.java
#	src/main/java/virtue/VirtueTaskList.java

sarjinius committed:
Finds coding style violations and resolves them

sarjinius committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Finds coding style violations and resolves them

sarjinius committed:
Incorporates a GUI into the chatbot.

sarjinius committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Incorporates a GUI into the chatbot.

sarjinius committed:
Adds assert checks to the code

sarjinius committed:
Refactors the code to improve code quality.

The sandwich and indent features are obsolete due to the new GUI.

The applyCommand method is better not without
a String output because it is public.

Let's
* remove the sandwich and indent methods in Virtue class
* remove its usages
* create a resultMessage field in Command class
* make the applyCommand method set the Command resultMessage

sarjinius committed:
Merge pull request #2 from sarjinius/branch-A-Assertions

Adds assert checks to the code

sarjinius committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Adds assert checks to the code

# Conflicts:
#	src/main/java/virtue/Virtue.java
#	src/main/java/virtue/VirtueTaskList.java

sarjinius committed:
Merge pull request #3 from sarjinius/branch-A-CodeQuality

Refactors the code to improve code quality.

sarjinius committed:
Merge branch 'master' of https://github.com/sarjinius/ip

* 'master' of https://github.com/sarjinius/ip:
  Refactors the code to improve code quality.

sarjinius committed:
Provide a way to perform commands on multiple tasks

The mark, unmark, and delete commands
can be applied to several tasks.

Adding this features makes it more convenient for the user.

Let's,
* update the getIndex method to accept multiple indices
* update the markTasksAsDone, markTasksAsNotDone, and deleteTasks
  methods to accept an array of indices

sarjinius committed:
Merge branch 'branch-C-MassOps'

* branch-C-MassOps:
  Provide a way to perform commands on multiple tasks

sarjinius committed:
Give credit for reused work

Several lines in the getIndices method were taken and modified from several websites.

Let's,
* add comments citing the sources for the code

sarjinius committed:
Add a representative screenshot to the docs folder

A good product screenshot showcases the product in its full glory.

Let's,
* produce a sample run with realistic data
* take a screenshot of the sample run
* add the screenshot to the docs folder

sarjinius committed:
Add a brief user guide

A user guide is needed for users to understand how to use
the chatbot's commands.

Let's add the command details and examples, together with
the Ui.png image, to the user guide.

sarjinius committed:
Provide appropriate messages to invalid inputs

When a user inputs an empty description or index,
or an index out of range, the user must be alerted.

Let's,
* check if a description is not entirely whitespace
* check if all indices are in the appropriate range
* alert the user if the conditions are not satisfied
